Mathlib/RingTheory/MvPowerSeries/Evaluation.lean

```diff
@@ -0,0 +1,396 @@
+
+Let `σ`, `R` `S` be types, with `CommRing R`, `CommRing S`.
+One assumes that `TopologicalRing R` and `UniformAddGroup R`,
+and that `S` is a complete and separated topological `R`-algebra,
+with `LinearTopology R`, which means there is a basis of neighborhoods of 0
+consisting of ideals.
+
+* `MvPowerSeries.eval₂` : Given `φ : R →+* S` and `a : σ → S`,
+this file defines an evaluation of `f : MvPowerSeries σ R`,
+that extends the evaluation of polynomials at `a`, by density,
+provided `φ` and `a` satisfy certain conditions of which
+the following lemmas assert the necessity
+
+* `Continuous.on_scalars` : The map `φ` is continuous
+* `Continuous.tendsto_apply_pow_zero_of_constantCoeff_zero` :
+  for all `s : σ`, `(a s) ^ n` tends to 0 when `n` tends to infinity
+* `Continuous.tendsto_apply_variables_zero_of_cofinite`:
+  when `a s` tends to  zero for the filter of cofinite subsets of `σ`.
+
+* `MvPowerSeries.eval₂_domain` : the `Prop`-valued structure
+that is required to evaluate power series
+
+* `MvPowerSeries.uniformContinuous_eval₂` : uniform continuity of the evaluation
+
+* `MvPowerSeries.continuous_eval₂` : continuity of the evaluation
+
+* `MvPowerSeries.aeval` : the evaluation map as an algebra map
+
+-/
+
+namespace MvPowerSeries
+
+open Topology
+
+open Filter MvPolynomial RingHom Set TopologicalSpace UniformSpace
+
+/- ## Necessary conditions -/
+
+section
+
+variable {σ : Type*}
+variable {R : Type*} [CommRing R] [TopologicalSpace R]
+variable {S : Type*} [CommRing S] [TopologicalSpace S]
+variable {φ : R →+* S}
+
+-- We endow MvPowerSeries σ R with the Pi topology
+open WithPiTopology
+
+/-- Families at which power series can be evaluated -/
+structure EvalDomain (a : σ → S) : Prop where
+  hpow : ∀ s, IsTopologicallyNilpotent (a s)
+  tendsto_zero : Tendsto a cofinite (𝓝 0)
+
+/-- The domain of evaluation of `MvPowerSeries`, as an ideal -/
+def EvalDomain_ideal [IsTopologicalRing S] [IsLinearTopology S S] :
+    Ideal (σ → S) where
+  carrier := setOf EvalDomain
+  add_mem' {a} {b} ha hb := {
+    hpow := fun s ↦ IsTopologicallyNilpotent.add (ha.hpow s) (hb.hpow s)
+    tendsto_zero := by
+      rw [← add_zero 0]
+      exact ha.tendsto_zero.add hb.tendsto_zero }
+  zero_mem' := {
+    hpow := fun s ↦ by
+      simp only [Pi.zero_apply]
+      apply tendsto_atTop_of_eventually_const (i₀ := 1)
+      intro i hi
+      rw [zero_pow (Nat.ne_zero_iff_zero_lt.mpr hi)]
+    tendsto_zero := tendsto_const_nhds }
+  smul_mem' c {x} hx := {
+    hpow := fun s ↦ by
+      simp only [IsTopologicallyNilpotent, Pi.smul_apply', smul_eq_mul, mul_pow]
+      exact IsLinearTopology.tendsto_mul_zero_of_right _ _ (hx.hpow s)
+    tendsto_zero := IsLinearTopology.tendsto_mul_zero_of_right _ _ hx.tendsto_zero }
+
+theorem EvalDomain.comp {a : σ → R} (ha : EvalDomain a) {ε : R →+* S} (hε : Continuous ε) :
+    EvalDomain (ε ∘ a) := by
+  apply EvalDomain.mk _ ((map_zero ε ▸ Continuous.tendsto' hε 0 (ε 0) rfl).comp ha.tendsto_zero)
```

Comment:
The goal of [Continuous.tendsto'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Basic.html#Continuous.tendsto') is precisely to avoid rewriting the value of `ε 0`
(otherwise [Continuous.tendsto](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Basic.html#Continuous.tendsto) does the job). 

```suggestion
  apply EvalDomain.mk _ ((Continuous.tendsto' hε 0 0 (map_zero ε)).comp ha.tendsto_zero)
```